### PR TITLE
Replace and delete files

### LINKED ISSUE #10: Support replacing files
Based on #9 it sounds like it might be a good idea to support patching files into the archives. For an initial proposal:

```ruby
act(
  archive_path: "example/Example.xcarchive",

  replace_files: {
    "ImageInRootOfApp.png" => "/path/to/source/image.png",
    "/ImageInRootOfArchive.png" => "/path/to/source/image.png"
  },
  remove_files: [
    "RedundantAppImage.png"
  ]
)
```

* `replace_files` feels better than `add_files` as it also implies existing files will be deleted, but is it clear enough that it can be used to add new files?
* The in-package paths should honour the same pattern as `plist_file` where the default is .app relative, and `/` prefix is archive-relative

This won't include patching the media catalog, as that is more complex and deserves it's own feature.

### PR DESCRIPTION
**Pull request by kimdv**
Fixes #10

### CODE REVIEW

**Comment by kimdv** [comment]
Well first attempt on learning Ruby! 
I will appreciate any feedback

**Comment by richardszalay** [comment]
Looks great! Can you also add a test into `ipa_archive`? I think you'll need to move swap over to use `archive.delete` instead of `replace` for deletes in order for it to work with IPAs

**Comment by kimdv** [comment]
Thanks for the feedback! 

I have added tests for `.ipa`.  And also fixed so they passes

File: lib/fastlane/plugin/act/helper/file_patcher.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      end
+
+      def self.remove(archive, file_values)
+        file_values.each do |file_to_delete|
+          UI.message("Deleting #{file_to_delete}")
+
+          relative_path = archive.app_path(file_to_delete)
+          local_path = archive.local_path(relative_path)
+
+          archive.extract(relative_path)
```
Comment (richardszalay) [inline lib/fastlane/plugin/act/helper/file_patcher.rb]: Extract isn't required here as `archive.delete` doesn't use it.

File: lib/fastlane/plugin/act/helper/file_patcher.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+          archive.replace(relative_path)
+        end
+      end
+
+      def self.remove(archive, file_values)
+        file_values.each do |file_to_delete|
+          UI.message("Deleting #{file_to_delete}")
+
+          relative_path = archive.app_path(file_to_delete)
+          local_path = archive.local_path(relative_path)
```
Comment (richardszalay) [inline lib/fastlane/plugin/act/helper/file_patcher.rb]: `local_path` isn't used, so you can remove this line.

File: spec/act_action_ipa_spec.rb [hunk 3]
```diff
@@ -174,6 +174,32 @@
           expect(result).to eql(["Blue29x29", "Blue40x40"])
         end
 
+        it 'replace files' do
```
Comment (richardszalay) [inline spec/act_action_ipa_spec.rb]: Can you create two separate tests for both app-relative and archive-relative paths work?

`replaces app-relative files` would replace "GoogleService-Info.plist" and validate against "Payload/Example.app/GoogleService-Info.plist"
`replaces archive-relative files` would replace "/Info.plist" (note forward slash prefix) and validate against "Info.plist"

File: spec/act_action_ipa_spec.rb [hunk 4]
```diff
@@ -174,6 +174,32 @@
...
+              "GoogleService-Info.plist" => "example/New-GoogleService-Info.plist"
+            }
+          )
+
+          result = invoke_plistbuddy("Print :TRACKING_ID", "Payload/Example.app/GoogleService-Info.plist")
+
+          expect(result).to eql("UA-123456789-12")
+        end
+
+        it 'delete files' do
```
Comment (richardszalay) [inline spec/act_action_ipa_spec.rb]: Can you create two separate tests for both app-relative and archive-relative paths work?

`delete app-relative files` would delete "GoogleService-Info.plist" and validate against "Payload/Example.app/GoogleService-Info.plist"
`deletes archive-relative files` would delete "/Info.plist" (note forward slash prefix) and validate against "Info.plist"

File: spec/act_action_xcarchive_spec.rb [hunk 5]
```diff
@@ -175,6 +175,32 @@
           expect(result).to eql(["Blue29x29", "Blue40x40"])
         end
 
+        it 'replace files' do
```
Comment (richardszalay) [inline spec/act_action_xcarchive_spec.rb]: See comments on ipa_archive tests

File: spec/act_action_xcarchive_spec.rb [hunk 6]
```diff
@@ -175,6 +175,32 @@
...
+              "GoogleService-Info.plist" => "example/New-GoogleService-Info.plist"
+            }
+          )
+
+          result = invoke_plistbuddy("Print :TRACKING_ID", "Products/Applications/Example.app/GoogleService-Info.plist")
+
+          expect(result).to eql("UA-123456789-12")
+        end
+
+        it 'delete files' do
```
Comment (richardszalay) [inline spec/act_action_xcarchive_spec.rb]: See comments on ipa_archive tests

**Comment by richardszalay** [comment]
Amazing work Kim, thanks so much for taking the time contribute.

I've added a few comments onto the PR, mainly around two additional test cases. If you wouldn't mind actioning those, I'm happy to merge and release.

I actually work from Windows so am limited to what I can do. However, TravisCI will run the tests and I can publish from within WSL on Windows.

File: spec/act_action_ipa_spec.rb (hunk 3 above)
Comment (richardszalay) [reply spec/act_action_ipa_spec.rb]: Actually can you move these into their own `context`

context 'replacing files'
it 'supports app-relative paths'
it 'supports archive-relative paths'

File: spec/act_action_ipa_spec.rb (hunk 4 above)
Comment (richardszalay) [reply spec/act_action_ipa_spec.rb]: Actually can you move these into their own context

context 'deleting files'
it 'supports app-relative paths'
it 'supports archive-relative paths'

File: spec/act_action_ipa_spec.rb (hunk 3 above)
Comment (richardszalay) [reply spec/act_action_ipa_spec.rb]: See here for how the use of `context` makes it easier to understand in the test results: https://travis-ci.org/richardszalay/fastlane-plugin-act/jobs/429428583

**Comment by kimdv** [comment]
@richardszalay I can't see any comments? :)

**Comment by richardszalay** [comment]
They should be inline in the file changes.

**Comment by kimdv** [comment]
Hmm 🤔 
I’m confused now... have you added comments here on GitHub or should I add comments ? 

Because I still can’t see any. Normally they will also appear on the timeline ? 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by richardszalay** [comment]
I've added review comments within the PR on GitHub, but not as additional commits.

**Comment by kimdv** [comment]
Nope, not available for me. Are they published?

**Comment by richardszalay** [comment]
Ah, didn't realise you needed to submit reviews 🤷‍♂️

**Comment by kimdv** [comment]
Hi @richardszalay !
Added some of you feedback! 

I don't get how the test case with `/Info.plist` should be?

Can you help me?

**Comment by richardszalay** [comment]
Sure, it's not really made explicit but app paths that start with a `/` don't get prefixed by the app paths.

For those tests, the replace_files key would be `/Info.plist` and the argument passed to `invoke_plistbuddy` would just be `Info.plist`

**Comment by kimdv** [comment]
So a test would look like this?

```
Fastlane::Actions::ActAction.run(
            archive_path: @ipa_file,
            remove_files: [
              "/Info.plist"
            ]
          )

          result = archive_contains("Payload/Example.app/Info.plist")

          expect(result).to be false
```

**Comment by richardszalay** [comment]
Close:

```ruby
result = archive_contains("Info.plist")
```

**Comment by kimdv** [comment]
Are you sure it works? 

`act_action_xcarchive_spec` fails
```
it 'deletes archive-relative files' do
          Fastlane::Actions::ActAction.run(
            archive_path: @archive_path,
            remove_files: [
              "/Info.plist"
            ]
          )

          result = archive_contains("Info.plist")

          expect(result).to be false
        end
```


`act_action_ipa_spec` fails
```
it 'deletes archive-relative files' do
          Fastlane::Actions::ActAction.run(
            archive_path: @ipa_file,
            remove_files: [
              "/Info.plist"
            ]
          )

          result = archive_contains("Info.plist")

          expect(result).to be false
        end
```

**Comment by kimdv** [comment]
And sorry for all the questions!

**Comment by richardszalay** [comment]
Hrm, it should! There are definitely plist-patching tests that use it. The functionality should be wrapped up in app_path / local_path, but maybe it's explicitly handled for plist patching?

**Comment by kimdv** [comment]
I have no idea. It sounds like it is handled by plist_buddy :)

**Comment by richardszalay** [comment]
Looks like it's handled by `ActAction` lines 22/23 but it is indeed specific to plist_path. 

Maybe it could be moved out to a new `ArchivePaths` helper class as `def self.expand(path)`? 

(`self.` assigns the function to the class rather than to an instance, similar to `static` in other languages - that way you can call it as `ArchivePaths.expand` without new'ing up an instance of `ArchivePaths`)

**Comment by kimdv** [comment]
I think I got it!

**Review by richardszalay — CHANGES_REQUESTED** [review]
Almost there!

File: spec/act_action_ipa_spec.rb (hunk 3 above)
Comment (richardszalay) [inline spec/act_action_ipa_spec.rb]: I think you've forgotten to remove these old tests

File: spec/act_action_ipa_spec.rb [hunk 7]
```diff
@@ -189,6 +215,49 @@
         end
       end
```
Comment (richardszalay) [inline spec/act_action_ipa_spec.rb:191]: archive-relative tests seem to be missing for 'replacing files'

File: spec/act_action_xcarchive_spec.rb [hunk 8]
```diff
@@ -190,6 +190,49 @@
         end
       end
 
+      context 'replacing files' do
```
Comment (richardszalay) [inline spec/act_action_xcarchive_spec.rb:193]: The archive-relative test is missing for replace here too

**Comment by richardszalay** [comment]
Awesome work Kim, I really appreciate your efforts.

I'll try to get a release out as soon as I can.

**Comment by kimdv** [comment]
Any idea of when the release will be available 🚀

**Comment by richardszalay** [comment]
I'll try within the next few days, but I'm on paternity leave with a newborn so my time is somewhat limited

**Comment by kimdv** [comment]
I see! 
Congratulations ! 🎉

**Comment by richardszalay** [comment]
Released as v1.2.0

**Pull request merged**